File: 206. Reverse Linked List/206.ReverseLinkedList.java

/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode reverseList(ListNode head) {
        return reverseListRecursive(head, null);
    }

    private ListNode reverseListRecursive(ListNode current, ListNode prev) {
        if (current == null) return prev; // Base case: end of list
        ListNode next = current.next; // Save next node
        current.next = prev; // Reverse the link
        return reverseListRecursive(next, current); // Recurse with next node and current as new prev
    }
}

File: 206. Reverse Linked List/206.ReverseLinkedList.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */

/**
 * Recursively reverse the linked list.
 * @param {ListNode} head
 * @return {ListNode}
 */
var reverseList = function (head, prev = null) {
  if (!head) return prev; // Base case: end of the list reached, return new head
  let next = head.next; // Save next node
  head.next = prev; // Reverse current node's pointer
  return reverseList(next, head); // Recurse on the next node
};
